### PR TITLE
React: Delete account functionality

### PR DESCRIPTION
**Pull request by cehali**
# Summary
Added delete account functionality and change component to work with it

Fixes: 0

## Checklist
- [x] Change is small and easy to review (please split big changes into multiple PRs)
- [x] Change is consistent with architecture
- [x] Change is consistent with test architecture
- [x] Change is consistent with naming conventions
- [x] New code is covered with tests
- [x] Tests related to old code are updated
- [x] Documentation is up to date with changes

### CODE REVIEW

File: universal-login-react/src/core/services/DeleteAccountService.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+export const getInputClassName = (inputError: boolean) => inputError ? 'delete-account-input-error' : 'delete-account-input';
+
+const doesAnyErrorExists = (errors: ErrorsType) => errors.usernameError || errors.verifyFieldError;
+
+export const deleteAccount = async (deployedWallet: DeployedWallet, inputs: InputsType, setErrors: (errors: ErrorsType) => void, onDeleteAccountClick: () => void) => {
+  const errors = checkInputsAgainstError(deployedWallet.name, inputs);
+  setErrors(errors);
+  if (!doesAnyErrorExists(errors)) {
+    const publicKey = utils.computeAddress(deployedWallet.privateKey);
```
Comment (Marik-D) [inline universal-login-react/src/core/services/DeleteAccountService.ts]: Make this as a method of `DeployedWallet`
Something like `removeSelf` or `disconnect`

File: universal-login-react/src/core/services/DeleteAccountService.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  const verifyFieldText = 'DELETE MY ACCOUNT';
+  return verifyFieldFromInput === verifyFieldText;
+};
+
+const checkInputsAgainstError = (usernameLogged: string, inputs: InputsType) => ({
+  usernameError: !validateUsername(inputs.username, usernameLogged),
+  verifyFieldError: !validateVerifyField(inputs.verifyField)
+});
+
+export const getInputClassName = (inputError: boolean) => inputError ? 'delete-account-input-error' : 'delete-account-input';
```
Comment (Marik-D) [inline universal-login-react/src/core/services/DeleteAccountService.ts]: This code deals with UI (HTML and React), it shouldn't be in services

File: universal-login-react/src/core/services/DeleteAccountService.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import {utils} from 'ethers';
+import {transactionDetails} from '../../core/constants/TransactionDetails';
+import {DeployedWallet} from '@universal-login/sdk';
+
+type ErrorsType = {
```
Comment (Marik-D) [inline universal-login-react/src/core/services/DeleteAccountService.ts]: Prefer interface
```suggestion
interface ErrorsType {
```

File: universal-login-react/src/core/services/DeleteAccountService.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  usernameError: boolean;
+  verifyFieldError: boolean;
+};
+
+type InputsType = {
```
Comment (Marik-D) [inline universal-login-react/src/core/services/DeleteAccountService.ts]:
```suggestion
interface InputsType {
```

File: universal-login-react/src/core/services/DeleteAccountService.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+const doesAnyErrorExists = (errors: ErrorsType) => errors.usernameError || errors.verifyFieldError;
+
+export const deleteAccount = async (deployedWallet: DeployedWallet, inputs: InputsType, setErrors: (errors: ErrorsType) => void, onDeleteAccountClick: () => void) => {
+  const errors = checkInputsAgainstError(deployedWallet.name, inputs);
+  setErrors(errors);
+  if (!doesAnyErrorExists(errors)) {
+    const publicKey = utils.computeAddress(deployedWallet.privateKey);
+    await deployedWallet.removeKey(publicKey, transactionDetails);
+    deployedWallet.sdk.stop();
```
Comment (Marik-D) [inline universal-login-react/src/core/services/DeleteAccountService.ts]: Don't stop the sdk

File: universal-login-react/src/core/services/DeleteAccountService.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    onDeleteAccountClick();
+  }
+};
```
Comment (Marik-D) [inline universal-login-react/src/core/services/DeleteAccountService.ts]: Consider inlining this logic in the component to reduce overall complexity

File: universal-login-react/src/core/services/DeleteAccountService.ts (hunk 4 above)
Comment (marekkirejczyk) [reply universal-login-react/src/core/services/DeleteAccountService.ts]: What is the difference? Why preference?

File: universal-login-react/src/core/services/DeleteAccountService.ts (hunk 2 above)
Comment (marekkirejczyk) [reply universal-login-react/src/core/services/DeleteAccountService.ts]: This is great example for the famous Application layer :) between UI and domain.

File: universal-login-wallet/src/ui/react/common/UserDropdown.tsx [hunk 7]
```diff
@@ -23,7 +19,7 @@ const UserDropdown = () => {
           <button onClick={isExpanded ?  collapseDropdown : expandDropdown} className="user-dropdown-btn" />
         </div>
         <div className={`user-dropdown-content ${isExpanded ? 'expanded' : ''}`}>
-          <button className="sign-out-btn" onClick={onDisconnectClick}>Disconnect</button>
+          <button className="sign-out-btn" onClick={() => disconnectFromWallet(walletService, router)}>Disconnect</button>
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/ui/react/common/UserDropdown.tsx]: We can remove the whole `UserDropdown`

File: universal-login-react/src/core/services/DeleteAccountService.ts (hunk 2 above)
Comment (Marik-D) [reply universal-login-react/src/core/services/DeleteAccountService.ts]: https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition

File: universal-login-react/package.json [hunk 8]
```diff
@@ -18,6 +18,7 @@
     "@ramp-network/ramp-instant-sdk": "^1.0.1",
     "@restless/sanitizers": "^0.2.2",
     "@universal-login/commons": "0.3.56",
+    "@universal-login/relayer": "0.3.56",
```
Comment (JustynaBroniszewska) [inline universal-login-react/package.json]: Move it to devDependencies!!!

File: universal-login-react/test/core/services/DeleteAccountService.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
+import chai, {expect} from 'chai';
+import sinon from 'sinon';
+import sinonChai from 'sinon-chai';
+import {deleteAccount} from '../../../src/core/services/DeleteAccountService';
+import {DeployedWallet, WalletService} from '@universal-login/sdk';
+import {Wallet} from 'ethers';
+import {getWallets, createMockProvider} from 'ethereum-waffle';
+import {setupDeployedWallet} from '../../helpers/setupDeploymentWallet';
+import Relayer from '@universal-login/relayer';
```
Comment (JustynaBroniszewska) [inline universal-login-react/test/core/services/DeleteAccountService.ts]: I believe this is `RelayerUnderTest`

Comment (cehali) [reply universal-login-react/test/core/services/DeleteAccountService.ts]: RelayerUnderTest doesn't create sdk and deployedWallet

Comment (JustynaBroniszewska) [reply universal-login-react/test/core/services/DeleteAccountService.ts]: A relayer that is returned in `setupSdk` is RelayerUnderTest type.

Comment (cehali) [reply universal-login-react/test/core/services/DeleteAccountService.ts]: Ok, sorry, I understood your comment wrongly

Comment (JustynaBroniszewska) [reply universal-login-react/test/core/services/DeleteAccountService.ts]: ok  👍

File: universal-login-wallet/test/integration/pages/DashboardPage.ts [hunk 10]
```diff
@@ -10,8 +10,16 @@ export default class DashboardPage {
     this.wrapper.update();
   }
 
-  disconnect() {
-    this.wrapper.find('.sign-out-btn').simulate('click');
+  areDeleteAccountInputsValid() {
```
Comment (rzadp) [inline universal-login-wallet/test/integration/pages/DashboardPage.ts]: Is this used anywhere? Besides, if it checks if valid, should be `! this.wrapper.exists('delete-account-input-error')` ?

File: universal-login-react/test/core/services/DeleteAccountService.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    expect(() => walletService.getDeployedWallet()).to.throw('Invalid state: expected deployed wallet');
+  });
+
+  it('dont delete account if inputs are invalid', async () => {
+    await deleteAccount(walletService, {username: 'test', verifyField: 'DELETE MY ACCOUNT'}, setErrors, onDeleteAccountClick);
+    expect(setErrors).to.be.calledOnce;
+    expect(onDeleteAccountClick).to.not.be.called;
+    expect(walletService.getDeployedWallet()).to.deep.eq(deployedWallet);
+  });
+
```
Comment (rzadp) [inline universal-login-react/test/core/services/DeleteAccountService.ts]: Add a test with wrong `verifyField`?

File: universal-login-react/src/core/services/DeleteAccountService.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+const checkInputsAgainstError = (usernameLogged: string, inputs: InputsType) => ({
+  usernameError: inputs.username !== usernameLogged,
+  verifyFieldError: inputs.verifyField !== 'DELETE MY ACCOUNT'
+});
+
+export const getInputClassName = (inputError: boolean) => inputError ? 'delete-account-input-error' : 'delete-account-input';
+
+const doesAnyErrorExists = (errors: ErrorsType) => errors.usernameError || errors.verifyFieldError;
+
+export const deleteAccount = async (walletService: WalletService, inputs: InputsType, setErrors: (errors: ErrorsType) => void, onDeleteAccountClick: () => void) => {
```
Comment (rzadp) [inline universal-login-react/src/core/services/DeleteAccountService.ts]: The `onDeleteAccountClick`, I'm not sure about the naming. Seems like it should be `onAccountDeleted`? I guess clicking 'delete account' happens before, which leds us here?

File: universal-login-wallet/test/integration/pages/DashboardPage.ts (hunk 10 above)
Comment (cehali) [reply universal-login-wallet/test/integration/pages/DashboardPage.ts]: 👍

File: universal-login-react/test/core/services/DeleteAccountService.ts (hunk 11 above)
Comment (cehali) [reply universal-login-react/test/core/services/DeleteAccountService.ts]: 👍

File: universal-login-react/src/core/services/DeleteAccountService.ts (hunk 12 above)
Comment (cehali) [reply universal-login-react/src/core/services/DeleteAccountService.ts]: 👍

**Comment by rzadp** [comment]
netlify rerun

File: universal-login-sdk/lib/api/DeployedWallet.ts [hunk 13]
```diff
@@ -45,6 +45,11 @@ export class DeployedWallet implements ApplicationWallet {
     return this.sdk.removeKey(this.contractAddress, key, this.privateKey, transactionDetails);
   }
 
+  async removeOwnKey(transactionDetails: Partial<Message>): Promise<Execution> {
+    const ownKey = utils.computeAddress(this.privateKey);
+    return this.sdk.removeKey(this.contractAddress, ownKey, this.privateKey, transactionDetails);
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/api/DeployedWallet.ts]: `return this.removeKey(ownKey, transactionDetails)`

File: universal-login-sdk/lib/core/services/WalletService.ts [hunk 14]
```diff
@@ -147,6 +147,13 @@ export class WalletService {
     };
   }
 
+  async removeWallet(transactionDetails: Partial<Message>) {
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/core/services/WalletService.ts]: rename `removeCurrentKey`

**Pull request merged**